File: crates/nil-core/src/player.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

mod manager;

use crate::infrastructure::storage::StorageCapacity;
use crate::resource::Resources;
use derive_more::{Display, From};
use serde::{Deserialize, Serialize};
use std::ops::Deref;
use std::sync::Arc;

pub use manager::PlayerManager;

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Player {
  id: PlayerId,
  status: PlayerStatus,
  resources: Resources,
}

impl Player {
  pub fn new(options: PlayerOptions) -> Self {
    Self {
      id: options.id,
      status: PlayerStatus::Active,
      resources: Resources::PLAYER.clone(),
    }
  }

  #[inline]
  pub fn id(&self) -> PlayerId {
    self.id.clone()
  }

  #[inline]
  pub fn status(&self) -> PlayerStatus {
    self.status
  }

  pub(crate) fn status_mut(&mut self) -> &mut PlayerStatus {
    &mut self.status
  }

  #[inline]
  pub fn resources(&self) -> &Resources {
    &self.resources
  }

  pub(crate) fn resources_mut(&mut self) -> &mut Resources {
    &mut self.resources
  }

  #[inline]
  pub fn is_active(&self) -> bool {
    matches!(self.status, PlayerStatus::Active)
  }

  #[inline]
  pub fn is_inactive(&self) -> bool {
    matches!(self.status, PlayerStatus::Inactive)
  }
}

#[derive(Debug, Display, From, PartialEq, Eq, Hash, Deserialize, Serialize)]
#[from(String, &str, Arc<str>, Box<str>)]
pub struct PlayerId(Arc<str>);

impl Clone for PlayerId {
  fn clone(&self) -> Self {
    Self(Arc::clone(&self.0))
  }
}

impl AsRef<str> for PlayerId {
  fn as_ref(&self) -> &str {
    &self.0
  }
}

impl Deref for PlayerId {
  type Target = str;

  fn deref(&self) -> &Self::Target {
    &self.0
  }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Deserialize, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum PlayerStatus {
  Active,
  Inactive,
}

#[derive(Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlayerOptions {
  pub id: PlayerId,
}

#[derive(Clone, Debug, Default, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct PlayerStorageCapacity {
  pub silo: StorageCapacity,
  pub warehouse: StorageCapacity,
}

File: ui/src/core/model/player.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import * as commands from '@/commands';
import { ResourcesImpl } from './resource';
import { CoordImpl } from '@/core/model/coord';
import type { PartialNullish } from '@tb-dev/utils';

export class PlayerImpl implements Player {
  public readonly id: string;
  public readonly resources: ResourcesImpl;
  public readonly status: PlayerStatus;
  public readonly coords: readonly CoordImpl[];

  private constructor(args: {
    id: string;
    resources: ResourcesImpl;
    status: PlayerStatus;
    coords: readonly CoordImpl[];
  }) {
    this.id = args.id;
    this.resources = args.resources;
    this.status = args.status;
    this.coords = args.coords;
  }

  public hasResources(resources: PartialNullish<Resources>) {
    return this.resources.has(resources);
  }

  public hasVillage(coord: Coord) {
    return this.coords.some((it) => it.is(coord));
  }

  public isActive() {
    return this.status === 'active';
  }

  public isInactive() {
    return this.status === 'inactive';
  }

  public static async load(id: PlayerId) {
    const [player, coords] = await Promise.all([
      commands.getPlayer(id),
      commands.getPlayerCoords(id),
    ]);

    return new PlayerImpl({
      id: player.id,
      resources: ResourcesImpl.create(player.resources),
      status: player.status,
      coords: coords.map((it) => CoordImpl.create(it)),
    });
  }
}
